sechub-shared-kernel/src/main/java/com/daimler/sechub/sharedkernel/resilience/ResilientActionExecutor.java

```diff
@@ -13,198 +15,243 @@
+        } else if (proposal instanceof FallthroughResilienceProposal) {
+            FallthroughResilienceProposal fallThroughProposal = (FallthroughResilienceProposal) proposal;
+            LOG.info("Fall through activated, will rethrow same exception {} milliseconds", fallThroughProposal.getMillisecondsForFallThrough());
+
+            context.forceFallThrough(fallThroughProposal);
+            throw e;
+        } else {
+            LOG.error("Returned proposal is not wellknown and so cannt be handled:{}", proposal.getClass().getName());
+            throw e;
+        }
+    }
+
+    private ResilienceProposal findFirstProposalFromConsultants(ResilienctActionContext context) {
+        ResilienceProposal proposal = null;
+        for (ResilienceConsultant consultant : consultants) {
+            proposal = consultant.consultFor(context);
+            if (proposal != null) {
+                break;
+            }
+        }
+        return proposal;
+    }
+
+    public void add(ResilienceConsultant consultant) {
+        Objects.requireNonNull(consultant, "consultant may not be null!");
+        consultants.add(consultant);
+    }
+
+    class ResilienctActionContext implements ResilienceContext {
+
+        private Exception currentError;
+        private int retriesCount;
+        private boolean retryNecessary;
+        private ResilienceCallback callBack;
+        private Map<String, Object> parameters = new HashMap<>(1);
+
+        public ResilienctActionContext(ResilienceCallback callBack) {
+            this.callBack = callBack;
+        }
+
+        public void prepareForNextCheck(Exception e) {
+            this.retryNecessary = false;
+            this.currentError = e;
+        }
+
+        public void forceFallThrough(FallthroughResilienceProposal proposal) {
+            this.retryNecessary = false;
+            fallThroughSupport.enable(currentError, proposal.getInfo(), proposal.getMillisecondsForFallThrough());
+        }
+
+        public boolean isRetryNecessary() {
+            return retryNecessary;
+        }
+
+        public void forceRetry() {
+            retryNecessary = true;
+            retriesCount++;
+        }
+
+        public void setCurrentError(Exception currentError) {
+            this.currentError = currentError;
+        }
+
+        @Override
+        public Exception getCurrentError() {
+            return currentError;
+        }
+
+        @Override
+        public int getAlreadyDoneRetries() {
+            return retriesCount;
+        }
+
+        /**
+         * @return callback or <code>null</code>
+         */
+        public ResilienceCallback getCallBack() {
+            return callBack;
+        }
+
+        @Override
+        public <V> void setValue(String key, V value) {
+            parameters.put(key, value);
+        }
+
+        @SuppressWarnings("unchecked")
+        public <V> V getValueOrNull(String key) {
+            Object v = parameters.get(key);
+            if (v == null) {
```

Comment:
I would rewrite this code. You initialize the result with `null` in any case.

I think the snipped would be easier to read by writing it this way:

~~~
V result = null;
Object parameter = parameters.get(key);
if (parameter != null) {
              try {
                result = (V) parameter;
            } catch (ClassCastException e) {
                LOG.error("Was not able to cast key:{} to wanted type. Was orignally: {}", key, v.getClass(), e);
            }
  }
  return result;
}
~~~

Reply:
Parameters ist just a map and a map has key and value. Also the method says "getValueOrNull" and not "getParameterOrNull"...

I am using guard closes - get out as fast as possible. I am not a fan of if chains at all - i was used to but not longer any more. 

one thing I should introduce is null handling for keys . I did use a hashmap and key null is not allowed. So I will prevent this by a small change.

Reply:
Will rename variable "paramters" to "map" - reduces iritations.